### ISSUE TITLE
Simplify user-assigned identities type

### ISSUE THREAD

**Issue by viveklak** [pulumi/pulumi-azure-native#812]
Currently assigning user-assigned identities requires jumping through hoops, e.g.:
In typescript
https://github.com/pulumi/pulumi-azure-native/blob/master/examples/keyvault/index.ts#L33

Assigning user assigned identities is pretty common use-case. 

A replacement of the following form would be much more usable `userAssignedIdentities: {[userIdentity.id]: {}}`.

**Comment by mikhailshilkov** [comment]
Or even `userAssignedIdentities: [userIdentity.id]`

**Comment by lukehoban** [comment]
> A replacement of the following form would be much more usable userAssignedIdentities: {[userIdentity.id]: {}}.

Or even better, understand this pattern and allow `userAssignedIdentities [userIdentity.id]`. (Presuming that there are no valid input properties allowed on the object provided here).

**Comment by clstokes** [comment]
For anyone that comes across this issue, here's how to achieve this in current implementation in Python. 

Specifically, the `id_to_dict` function specifically turns `<id>` into `{ <id>: {} }` and then is called in the form `managed_identity.id.apply(lambda val: id_to_dict(val))`.

```py
from typing import Any, Mapping

from pulumi_azure_native import managedidentity
from pulumi_azure_native import storage
from pulumi_azure_native import resources

resource_group = resources.ResourceGroup('resource_group')

managed_identity = managedidentity.UserAssignedIdentity("main",
    resource_group_name=resource_group.name,
)

# Turn `<id>` into `{ <id>: {} }`
def id_to_dict(id_output) -> Mapping[str, Any]:
    my_dict = {}
    my_dict[id_output] = {}
    return my_dict

account = storage.StorageAccount('sa',
    resource_group_name=resource_group.name,
    sku=storage.SkuArgs(
        name=storage.SkuName.STANDARD_LRS,
    ),
    kind=storage.Kind.STORAGE_V2,
    identity=storage.IdentityArgs(
        type="UserAssigned",
        user_assigned_identities=managed_identity.id.apply(lambda val: id_to_dict(val)),
    ))
```

**Comment by Aaronontheweb** [comment]
I have ran into this sample problem with C# and haven't figured out how to make this work (assigning to a Function App)

```csharp
Identity = new ManagedServiceIdentityArgs()
            {
                Type = Pulumi.AzureNative.Web.ManagedServiceIdentityType.UserAssigned,
                UserAssignedIdentities = userAssignedIdentity.Id.Apply(id =>
                {
                    var dict = new Dictionary<string, object> { [id] = new { } };
                    return dict;
                })
            },
```

With this code I get back:

```
System.InvalidOperationException: <>f__AnonymousType0 is not a supported argument type.
```

Remove the `new` keyword:

```csharp
Identity = new ManagedServiceIdentityArgs()
            {
                Type = Pulumi.AzureNative.Web.ManagedServiceIdentityType.UserAssigned,
                UserAssignedIdentities = userAssignedIdentity.Id.Apply(id =>
                {
                    var dict = new Dictionary<string, object> { [id] = { } };
                    return dict;
                })
            },
```

And I get this back instead:

```
error: Code="MissingIdentityIds" Message="The identity ids must not be null or empty for 'UserAssigned' identity type."
```

**Comment by mikhailshilkov** [comment]
@Aaronontheweb Apologies for the tricky API here.

I have this snippet saved - it worked for a managed AKS cluster:

```cs
Identity = new AzureNative.ContainerService.Inputs.ManagedClusterIdentityArgs
{
    Type = AzureNative.ContainerService.ResourceIdentityType.UserAssigned,
    UserAssignedIdentities =
    {
        { "/subscriptions/subid/resourcegroups/rgname/providers/Microsoft.ManagedIdentity/userAssignedIdentities/test", "" }
    }
}
```

If you share a full code sample for the function app, I can try it locally and see what goes wrong.

**Comment by mikhailshilkov** [comment]
I edited the example above - I had a bug in the original.

**Comment by Aaronontheweb** [comment]
```csharp
Identity = new ManagedServiceIdentityArgs()
            {
                Type = Pulumi.AzureNative.Web.ManagedServiceIdentityType.UserAssigned,
                UserAssignedIdentities = userAssignedIdentity.Id.Apply(id =>
                {
                    var im = new InputMap<object>()
                    {
                        { id, new Dictionary<string, object>() }
                    };
                    return im;
                })

            },
```

Can't compile:

**Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0029	Cannot implicitly convert type 'Pulumi.Output<Pulumi.InputMap<object>>' to 'Pulumi.InputMap<object>'	Petabridge.Collector.Infra	D:\Repositories\Petabridge\kpi-dashboard\src\Petabridge.Collector.Infra\KpiAzureFunctionStack.cs	170	Active
**

**Comment by mikhailshilkov** [comment]
Why `InputMap`? This should compile:

```cs
Identity = new ManagedClusterIdentityArgs
{
    Type = ResourceIdentityType.UserAssigned,
    UserAssignedIdentities = userAssignedIdentity.Id.Apply(id =>
    {
        var im = new Dictionary<string, object>
        {
            { id, new Dictionary<string, object>() }
        };
        return im;
    })
},
```

**Comment by davejhahn** [comment]
@clstokes thanks for sharing that method--I wasted a lot of time trying to get this to work and your method worked.

**Comment by larsmaesmosadexehealth** [comment]
THanks for all the examples, because of this I am able to get this to work for one userassigned identity. Now i have this use case of assigning multiple identities.. how am I supposed to handle this. I have tried some thing with adding identities to a dictionary that is instaniated outside the Apply, but this does not seem to work.. any tips would be more than welcome

**Comment by mikhailshilkov** [comment]
@larsmaesmosadexehealth Sorry for a slow response. Which language are you using and which base example worked for you? Based on your question, I expect you to need to use `All` or `Tuple` to combine multiple outputs and then build a dictionary is the `Apply`. Something like

```cs
UserAssignedIdentities = Output.Tuple(userAssignedIdentity1.Id, userAssignedIdentity2.Id).Apply(ids =>
{
        var im = new Dictionary<string, object>
        {
            { ids.Item1, new Dictionary<string, object>() },
            { ids.Item2, new Dictionary<string, object>() }
        };
        return im;
})
```

**Comment by pierskarsenbarg** [comment]
I just wanted to add a typescript version of Cameron's code for when I come searching for this next time:

```typescript
const getId = (id: string) => {
    const dict: {[key: string]: object} = {};
    dict[id] = {};
    return dict; 
}

const kubernetesCluster = new containerservice.ManagedCluster('cluster', {
    location: resourceGroup.location,
    resourceGroupName: resourceGroup.name,
    identity: {
        type: "UserAssigned",
        userAssignedIdentities: clusterIdentity.id.apply(id => getId(id))
    }
});
```

**Comment by TechnicallyJoe** [comment]
> I just wanted to add a typescript version of Cameron's code for when I come searching for this next time:
> 
> ```ts
> const getId = (id: string) => {
>     const dict: {[key: string]: object} = {};
>     dict[id] = {};
>     return dict; 
> }
> 
> const kubernetesCluster = new containerservice.ManagedCluster('cluster', {
>     location: resourceGroup.location,
>     resourceGroupName: resourceGroup.name,
>     identity: {
>         type: "UserAssigned",
>         userAssignedIdentities: clusterIdentity.id.apply(id => getId(id))
>     }
> });
> ```

Think you could provide us with a Go version of this as well? - I can't for the life of me get this to work

**Comment by machscott** [comment]
> For anyone that comes across this issue, here's how to achieve this in current implementation in Python.
> 
> Specifically, the `id_to_dict` function specifically turns `<id>` into `{ <id>: {} }` and then is called in the form `managed_identity.id.apply(id_to_dict)`.
> 
> ```python
> from typing import Any, Mapping
> 
> from pulumi_azure_native import managedidentity
> from pulumi_azure_native import storage
> from pulumi_azure_native import resources
> 
> resource_group = resources.ResourceGroup('resource_group')
> 
> managed_identity = managedidentity.UserAssignedIdentity("main",
>     resource_group_name=resource_group.name,
> )
> 
> # Turn `<id>` into `{ <id>: {} }`
> def id_to_dict(id_output) -> Mapping[str, Any]:
>     my_dict = {}
>     my_dict[id_output] = {}
>     return my_dict
> 
> account = storage.StorageAccount('sa',
>     resource_group_name=resource_group.name,
>     sku=storage.SkuArgs(
>         name=storage.SkuName.STANDARD_LRS,
>     ),
>     kind=storage.Kind.STORAGE_V2,
>     identity=storage.IdentityArgs(
>         type="UserAssigned",
>         user_assigned_identities=managed_identity.id.apply(id_to_dict),
>     ))
> ```

Here is a one line version for Python that I have working right now.
`
"user_assigned_identities":managed_identity.id.apply(lambda name: {name: {}})
`

**Comment by cubikca** [comment]
This does work for `InputMap<object>`, but I guess because `Dictionary<string, object>` serializes to `{}`. With respect, I don't think this is a solution. I'm going to echo the thoughts above that there are far too many hoops to jump through for what is probably a pretty common use case. I'll also point out that not all Identity configurations use `InputMap<object>`, as in the following example:

```csharp
          language = new Account(name, new AccountArgs
          {
              Kind = "TextAnalytics",
              AccountName = name,
              Identity = new AzureNative.CognitiveServices.Inputs.IdentityArgs
              {
                  Type = AzureNative.CognitiveServices.IdentityType.UserAssigned,
                  UserAssignedIdentities =
                  {
                      // [id] = ???
                      // this is what the API documentation seems to imply would work:
                      [id] = new AzureNative.CognitiveServices.Inputs.UserAssignedIdentityArgs
                      {
                          PrincipalId = languageIdentity.PrincipalId
                      }
                  }
              },
              ResourceGroupName = _cognitiveRg.Name,
              Sku = new AzureNative.CognitiveServices.Inputs.SkuArgs
              {
                  Name = "S"
              },
              Properties = new AzureNative.CognitiveServices.Inputs.CognitiveServicesAccountPropertiesArgs
              {
                  PublicNetworkAccess = AzureNative.CognitiveServices.PublicNetworkAccess.Disabled,
                  CustomSubDomainName = name,
                  UserOwnedStorage = new AzureNative.CognitiveServices.Inputs.UserOwnedStorageArgs
                  {
                     ResourceId = _coreRef.GetOutput("StorageAccountId").Apply(sa => Convert.ToString(sa) ?? string.Empty)
                  }
              }
          });
```
We cannot use the same trick of assigning `Dictionary<string, object>`, because it won't compile. And this code doesn't work.

**Comment by mikhailshilkov** [comment]
@cubikca and anyone having this problem: you can definitely make it work in C# but it may require some `Apply` machinery. You haven't shown all the types of the variables, but assuming `id` is `Output<string>` defined somewhere above your snippet, if would be

```cs
UserAssignedIdentities = id.Apply(id =>
    new Dictionary<string, AzureNative.CognitiveServices.Inputs.UserAssignedIdentityArgs>
    {
        [id] = new() { PrincipalId = languageIdentity.PrincipalId }
    })
```

**Comment by cubikca** [comment]
I'm pretty sure I tried this combination and got an error about the JSON being incorrect, as new() { PrincipalId = ... } doesn't serialize to either null or `{}`. If I can find some time to get back to this, I'll have a closer look. For now, though, I think unwieldly is too kind a word.

**Comment by eff917** [comment]
I have run into tha same issue, not being able to assign an identity to a managed cluster in python. in typescript it does work, howerver in python I get the following error:
```azure-native:containerservice:ManagedCluster (efftest-k8s):
    error: Code="UnmarshalError" Message="Invalid request body. Converting request body to a managed cluster encountered error: Properties.properties.identityProfile accept type v20210301.UserAssignedIdentity, not type string."```
The code that produces this is the following:
```
    ...
    userIdentity=aksUserIdentity.id.apply(lambda name: {name: {}})
    managedCluster = pulumi_azure_native.containerservice.ManagedCluster(
        ....
        identity=pulumi_azure_native.containerservice.ManagedClusterIdentityArgs(
            type=pulumi_azure_native.containerservice.ResourceIdentityType.USER_ASSIGNED,
            user_assigned_identities=userIdentity
        ),
```

**Comment by danielrbradley** [comment]
This is now merged into the `v2.0-pre` branch ready for the 2.0 release. This is available for testing in the 2.0.0-alpha.2 release

**Comment by mswietlicki** [comment]
For C# developers:

```csharp
public static InputMap<object> GetUserAssignedIdentities(params Input<string>[] ids) =>
    Output.All(ids).Apply(ids => ids.ToDictionary(id => id, _ => (object)new Dictionary<string, object>()));
```
